### PR TITLE
LLM_EXTRACT_TEXT implementation

### PR DESCRIPTION
**Pull request by charleschile**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #18664 

## What this PR does / why we need it:

As part of our document LLM support, we are introducing the `LLM_EXTRACT_TEXT` function. This function extracts text from PDF files and writes the extracted text to a specified text file, extractor type can be specified by the third argument.

**Usage:** `llm_extract_text(<input PDF datalink>, <output text file datalink>, <extractor type string>);`

**Return Value:** A boolean indicating whether the extraction and writing process was successful.

**Note:**

- Both the input and output paths must be absolute paths.

Example SQL:

```sql
select llm_extract_text(cast('file:///Users/charles/Desktop/codes/matrixone/matrixone/test/distributed/resources/llm_test/extract_text/example.pdf?offset=0&size=4' as datalink), cast('file:///Users/charles/Desktop/codes/matrixone/matrixone/test/distributed/resources/llm_test/extract_text/example.txt' as datalink), "pdf");
```



Example return:

```shell
mysql> select llm_extract_text(cast('file:///Users/charles/Desktop/codes/matrixone/matrixone/test/distributed/resources/llm_test/extract_text/example.pdf?offset=0&size=4' as datalink), cast('file:///Users/charles/Desktop/codes/matrixone/matrixone/test/distributed/resources/llm_test/extract_text/example.txt' as datalink), "pdf");
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| llm_extract_text(cast(file:///Users/charles/Desktop/codes/matrixone/matrixone/test/distributed/resources/llm_test/extract_text/example.pdf?offset=0&size=4 as datalink), cast(file:///Users/charles/Desktop/codes/matrixone/matrixone/test/distributed/resources/llm_test/extract_text/example.txt as datalink), pdf)   |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| true                                                                                                                                                                                                                                                                                                                    |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.10 sec)

```

### CODE REVIEW

**Review by fengttt — CHANGES_REQUESTED** [review]
How do you know you are extracting a pdf file?     How about add a const "file type" or "extractor type" arg to the function? 

Extractor type arg might be better, if we happen to have many different kind of extractors for same file type (or a extractor for many file types).

**Review by fengttt — CHANGES_REQUESTED** [review]
Please add a test.

**Comment by charleschile** [comment]
> How do you know you are extracting a pdf file? How about add a const "file type" or "extractor type" arg to the function?
> 
> Extractor type arg might be better, if we happen to have many different kind of extractors for same file type (or a extractor for many file types).

Extractor type arg added

**Comment by charleschile** [comment]
> Please add a test.

Unit and BVT tests have been added.

File: test/distributed/cases/function/func_llm_extract_file.sql [hunk 1]
```diff
@@ -0,0 +1,2 @@
+select llm_extract_text(cast('file://$resources/llm_test/extract_text/MODocs1.pdf?offset=0&size=4' as datalink), cast('file://$resources/llm_test/extract_text/MODocs1.txt' as datalink), "pdf");
```
Comment (fengttt) [inline test/distributed/cases/function/func_llm_extract_file.sql:1]: Why do we need offset and size here?

**Review by fengttt — CHANGES_REQUESTED** [review]
Please add result verification in the bvt test.

File: test/distributed/cases/function/func_llm_extract_file.sql (hunk 1 above)
Comment (charleschile) [reply test/distributed/cases/function/func_llm_extract_file.sql:1]: I included them here to demonstrate that this format of datalink is recognized by the function. 
Offset and size are not utilized in the llm_extract_text function and will not affect the outcome.
Shall I delete them?

File: pkg/sql/plan/function/func_llm.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+func extractTextFromPdfAndWriteToFile(pdfPath string, txtPath string, proc *process.Process) (bool, error) {
+	// read PDF to string
+	content, err := readPdfToString(pdfPath)
+	if err != nil {
+		return false, moerr.NewInvalidInputNoCtxf("Invalid PDF input.")
+	}
+
+	// file service and write file
+	ctx := context.TODO()
```
Comment (cpegeric) [inline pkg/sql/plan/function/func_llm.go:98]: try to get the context from proc.  Only use context.TODO() when you have no choice.

File: pkg/sql/plan/function/func_llm.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+	fs, readPath, err := fileservice.GetForETL(ctx, proc.Base.FileService, txtPath)
+
+	// delete the file if txt file exist because Write() only works when a file does not exist
+	_, err = fs.StatFile(ctx, readPath)
```
Comment (cpegeric) [inline pkg/sql/plan/function/func_llm.go:102]: Just error out if file exists.  You may delete the customer file in the cloud if you use stage URL.

File: pkg/sql/plan/function/func_llm.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+func isSameSentence(current, last pdf.Text) bool {
+	return strings.TrimSpace(current.S) != "" &&
+		last.Font == current.Font &&
+		last.FontSize == current.FontSize &&
+		last.X == current.X &&
+		last.Y == current.Y
+}
+
+func readPdfToString(path string) (string, error) {
+	f, r, err := pdf.Open(path)
```
Comment (cpegeric) [inline pkg/sql/plan/function/func_llm.go:142]: Can you use fileservice to open a pdf file?

**Review by cpegeric — CHANGES_REQUESTED** [review]

File: test/distributed/cases/function/func_llm_extract_file.sql (hunk 1 above)
Comment (fengttt) [reply test/distributed/cases/function/func_llm_extract_file.sql:1]: Yeah, just remove those.

**Review by fengttt — APPROVED** [review]

**Review by qingxinhome — APPROVED** [review]

**Review by m-schen — APPROVED** [review]

**Review by daviszhen — APPROVED** [review]

**Review by XuPeng-SH — APPROVED** [review]

**Review by aressu1985 — APPROVED** [review]

**Pull request merged**